torchrl/modules/models/multiagent.py

```diff
@@ -230,6 +230,221 @@ def forward(self, *inputs: Tuple[torch.Tensor]) -> torch.Tensor:
+        False
+    """
+
+    def __init__(
+        self,
+        n_agents: int,
+        centralised: bool,
+        share_params: bool,
+        device: Optional[DEVICE_TYPING] = None,
+        num_cells: Optional[Sequence[int]] = None,
+        kernel_sizes: Union[Sequence[Union[int, Sequence[int]]], int] = 5,
+        strides: Union[Sequence, int] = 2,
+        paddings: Union[Sequence, int] = 0,
+        activation_class: Type[nn.Module] = nn.ELU,
+        **kwargs,
+    ):
+        super().__init__()
+
+        self.n_agents = n_agents
+        self.centralised = centralised
+        self.share_params = share_params
+
+        self.agent_networks = nn.ModuleList(
+            [
+                ConvNet(
+                    num_cells=num_cells,
+                    kernel_sizes=kernel_sizes,
+                    strides=strides,
+                    paddings=paddings,
+                    activation_class=activation_class,
+                    device=device,
+                    **kwargs,
+                )
+                for _ in range(self.n_agents if not self.share_params else 1)
+            ]
+        )
+
+    def forward(self, inputs: torch.Tensor):
+        assert (
+            len(inputs.shape) >= 4
+        ), """Multi-agent network expects" (*batch_size, agent_index, x, y, channels)"""
+        assert (
+            inputs.shape[-4] == self.n_agents
+        ), f"""Multi-agent network expects {self.n_agents} but got {inputs.shape[-4]}"""
+        # If the model is centralized, agents have full observability
+        if self.centralised:
+            shape = (
+                *inputs.shape[:-4],
+                self.n_agents * inputs.shape[-3],
+                inputs.shape[-2],
+                inputs.shape[-1],
+            )
+            inputs = torch.reshape(inputs, shape)
+
+        # If the parameters are not shared, each agent has its own network
+        if not self.share_params:
+            if self.centralised:
+                output = torch.stack(
+                    [net(inputs) for net in self.agent_networks], dim=-2
+                )
+            else:
+                output = torch.stack(
+                    [
+                        net(inputs[..., i, :, :, :])
+                        for i, net in enumerate(self.agent_networks)
+                    ],
+                    dim=-2,
+                )
+        else:
+            output = self.agent_networks[0](inputs)
+            if self.centralised:
+                # If the parameters are shared, and it is centralised all agents will have the same output.
+                # We expand it to maintain the agent dimension, but values will be the same for all agents
+                n_agent_outputs = output.shape[-1]
+                output = (
+                    output.view(*output.shape[:-1], n_agent_outputs)
+                    .unsqueeze(-2)
+                    .expand(*output.shape[:-1], self.n_agents, n_agent_outputs)
+                )
```

Comment:
we can maybe split that in mode lines if we need debugging?

Reply:
Not sure how adjusting this would help - each function call is already on an individual line

Reply:
We can't really put breakpoints on this I think.
And the error stack won't be super duper clear.

From a performance perspective I don't think python is considerably faster with multi-line code like this.

```suggestion
                output = output.view(*output.shape[:-1], n_agent_outputs)
                output = output.unsqueeze(-2)
                output = output.expand(*output.shape[:-1], self.n_agents, n_agent_outputs)
```